collections-lists/FastestWayToCheckIfListIsOrdered/FastestWayToCheckIfListIsOrdered/ListOrderValidator.cs

```diff
@@ -0,0 +1,202 @@
+﻿using System.Runtime.InteropServices;
+
+namespace FastestWayToCheckIfListIsOrdered;
+
+public class ListOrderValidator
+{
+	public static bool IsOrderedUsingForLoop<T>(IList<T> list, IComparer<T>? comparer = default)
+	{
+		if (list.Count <= 1) return true;
+
+		comparer ??= Comparer<T>.Default;
+		var length = list.Count;
+		for (var i = 1; i < length; i++)
+		{
+			if (comparer.Compare(list[i - 1], list[i]) > 0) return false;
+		}
+
+		return true;
+	}
+
+	public static bool IsOrderedUsingArraySort<T>(IList<T> list, T[] array, IComparer<T>? comparer = default)
+	{
+		if (list.Count <= 1) return true;
+
+		comparer ??= Comparer<T>.Default;
+
+		var length = list.Count;
+
+		list.CopyTo(array, 0);
+		Array.Sort(array, 0, length, comparer);
+
+		for (var i = 0; i < length; i++)
+		{
+			if (comparer.Compare(list[i], array[i]) != 0) return false;
+		}
+
+		return true;
+	}
+
+	public static bool IsOrderedUsingSpans<T>(List<T> list, IComparer<T>? comparer = default)
+	{
+		comparer ??= Comparer<T>.Default;
+		var span = CollectionsMarshal.AsSpan(list);
+
+		var length = span.Length;
+		for (var i = 1; i < length; i++)
+		{
+			if (comparer.Compare(span[i - 1], span[i]) > 0) return false;
+		}
+
+		return true;
+	}
+
+	public static bool IsOrderedUsingEnumerator<T>(IList<T> list, IComparer<T>? comparer = default)
+	{
+		if (list.Count <= 1) return true;
+
+		comparer ??= Comparer<T>.Default;
+		var previous = list[0];
+
+		using var enumerator = list.GetEnumerator();
+		enumerator.MoveNext();
+		while (enumerator.MoveNext())
+		{
+			var current = enumerator.Current;
+			if (comparer.Compare(previous, current) > 0) return false;
+			previous = current;
+		}
+
+		return true;
+	}
+
+	public static bool IsOrderedUsingLinqWithSequenceEqual<T>(IList<T> list, IComparer<T>? comparer = default,
+		IEqualityComparer<T>? equalityComparer = default)
+	{
+		if (list.Count <= 1) return true;
+
+		comparer ??= Comparer<T>.Default;
+		equalityComparer ??= EqualityComparer<T>.Default;
+
+		var orderedList = list.Order(comparer);
+		return list.SequenceEqual(orderedList, equalityComparer);
+	}
+
+	public static bool IsOrderedUsingLinqWithOrder<T>(IList<T> list, IComparer<T>? comparer = default)
+	{
+		if (list.Count <= 1) return true;
+
+		comparer ??= Comparer<T>.Default;
+		var orderedList = list.Order(comparer);
+
+		using var enumerator = orderedList.GetEnumerator();
+		enumerator.MoveNext();
+
+		var length = list.Count;
+		for (var i = 0; i < length; i++)
+		{
+			if (comparer.Compare(list[i], enumerator.Current) != 0) return false;
+			enumerator.MoveNext();
+		}
```

Comment:
Maybe the enumerator is faster than a foreach here, not sure, but if they are basically the same performance, we should favor the foreach since the code will be shorter and easy to follow.